src/transformers/models/mamba/modeling_mamba.py

```diff
@@ -81,14 +82,58 @@ def __init__(
         ssm_state_size = config.state_size
         conv_kernel_size = config.conv_kernel
 
-        self.conv_states = {
-            i: torch.zeros(batch_size, intermediate_size, conv_kernel_size, device=device, dtype=dtype)
-            for i in range(config.num_hidden_layers)
-        }
-        self.ssm_states = {
-            i: torch.zeros(batch_size, intermediate_size, ssm_state_size, device=device, dtype=dtype)
-            for i in range(config.num_hidden_layers)
-        }
+        self.conv_state_cache: List[torch.Tensor] = []
+        self.ssm_state_cache: List[torch.Tensor] = []
+
+        self.is_cache_initialized = [False for _ in range(config.num_hidden_layers)]
+
+        for i in range(config.num_hidden_layers):
+            new_layer_conv_cache = torch.zeros(
+                batch_size, intermediate_size, conv_kernel_size, device=device, dtype=dtype
+            )
+            new_layer_ssm_cache = torch.zeros(
+                batch_size, intermediate_size, ssm_state_size, device=device, dtype=dtype
+            )
+            torch._dynamo.mark_static_address(new_layer_conv_cache)
+            torch._dynamo.mark_static_address(new_layer_ssm_cache)
+            self.conv_state_cache.append(new_layer_conv_cache)
+            self.ssm_state_cache.append(new_layer_ssm_cache)
+
+    def update_conv_state(self, layer_idx: int, new_conv_state: torch.Tensor) -> torch.Tensor:
+        conv_state = self.conv_state_cache[layer_idx]
+        if not self.is_initialized(layer_idx):
+            conv_state += new_conv_state.to(conv_state.device)
+            return conv_state
+
+        # decoding phase
+        conv_state = conv_state.roll(shifts=-1, dims=-1)
+        conv_state[:, :, -1] = new_conv_state[:, :, 0]
+        self.conv_state_cache[layer_idx].zero_()
+        self.conv_state_cache[layer_idx] += conv_state
+        return self.conv_state_cache[layer_idx]
+
+    def get_conv_sate(self, layer_idx):
+        return self.conv_state_cache[layer_idx]
+
+    def update_ssm_state(self, layer_idx: int, new_ssm_state: torch.Tensor):
+        ssm_state = self.ssm_state_cache[layer_idx]
+        ssm_state.zero_()
+        ssm_state += new_ssm_state.to(ssm_state.device)
+        self.is_cache_initialized[layer_idx] = True
+        return ssm_state
+
+    def get_ssm_sate(self, layer_idx):
+        return self.ssm_state_cache[layer_idx]
+
+    def is_initialized(self, layer_idx):
+        return self.is_cache_initialized[layer_idx]
```

Comment:
I think this is fine just like we need a flag in whisper, here `cache_positions` is not so meaningful because we always know how to update and get the cache even if `cache_positions` is not passed

Reply:
Theoretically yes, but this is adding some complexity, which is not needed in the cache API. Checking the cache positions is more reliable, and is what we want to go with. 
- you don't have to reset and set another tensor
which is also a win

Let's just use the cache positions